README.md

<!-- README.md is generated from README.Rmd. Please edit that file and knit -->

# wateRinfo <img src="man/figures/logo.png" align="right" alt="" width="120">

[![Build
Status](https://travis-ci.org/ropensci/wateRinfo.svg?branch=addci)](https://travis-ci.org/ropensci/wateRinfo)
[![AppVeyor Build
status](https://ci.appveyor.com/api/projects/status/5ia9va0jyskbvhev/branch/master?svg=true)](https://ci.appveyor.com/project/ropensci/waterinfo/branch/master)
[![Coverage
Status](https://coveralls.io/repos/github/ropensci/wateRinfo/badge.svg)](https://coveralls.io/r/ropensci/wateRinfo?branch=master)

The goal of wateRinfo is to facilitate access to the variety of
environmental water-related data about Flanders (Belgium) available on
[waterinfo.be](https://www.waterinfo.be/) by providing an R-interface to
download time series data.

An
[API](https://www.waterinfo.be/download/9f5ee0c9-dafa-46de-958b-7cac46eb8c23?dl=0)
is provided by waterinfo.be to request time series data. However, this
still requires the proper composition of the URL with the identification
codes (`Timeseriesgroup_id` and `ts_id`) as used by the system itself.
To facilitate the user in searching for stations and variables and
consequently download the data of interest, this package provides some
general R functions to download the `waterinfo.be` data.

## Installation

You can install wateRinfo from GitHub with:

``` r
# install.packages("devtools")
devtools::install_github("ropensci/wateRinfo")
```

When succesfull, load it as usual:

``` r
library(wateRinfo)
```

## Example

For a number of supported variables
([documented](https://www.waterinfo.be/download/9f5ee0c9-dafa-46de-958b-7cac46eb8c23?dl=0)
by VMM), the stations providing time series data for a given variable
can be listed with the command `get_stations`.

If you want to know the supported variables, ask for the supported
variables:

``` r
supported_variables("en")
#>              variable_en
#> 1              discharge
#> 6        soil_saturation
#> 7          soil_moisture
#> 8  dew_point_temperature
#> 9     ground_temperature
#> 10           ground_heat
#> 11            irradiance
#> 12          air_pressure
#> 13 air_temperature_175cm
#> 14              rainfall
#> 20     relative_humidity
#> 21  evaporation_monteith
#> 25    evaporation_penman
#> 29        water_velocity
#> 34           water_level
#> 39     water_temperature
#> 40        wind_direction
#> 41            wind_speed
```

Listing the available air\_pressure stations:

``` r
get_stations("air_pressure")
#>      ts_id station_latitude station_longitude station_id station_no
#> 1 78124042         51.20300          5.439589      12213   ME11_002
#> 2 78039042         51.24379          4.266912      12208   ME04_001
#> 3 78005042         51.02263          2.970584      12206   ME01_003
#> 4 78073042         50.88663          4.094898      12210   ME07_006
#> 5 78107042         51.16224          4.845708      12212   ME10_011
#> 6 78022042         51.27226          3.728299      12207   ME03_017
#> 7 78090042         50.73795          5.141976      12211   ME09_012
#> 8 78056042         50.86149          3.411318      12209   ME05_019
#>              station_name stationparameter_name parametertype_name
#> 1             Overpelt_ME                    Pa                 Pa
#> 2              Melsele_ME                    Pa                 Pa
#> 3               Zarren_ME                    Pa                 Pa
#> 4           Liedekerke_ME                    BP                 Pa
#> 5            Herentals_ME                    Pa                 Pa
#> 6            Boekhoute_ME                    Pa                 Pa
#> 7 Niel-bij-St.-Truiden_ME                    Pa                 Pa
#> 8              Waregem_ME                    Pa                 Pa
#>   ts_unitsymbol dataprovider
#> 1           hPa          VMM
#> 2           hPa          VMM
#> 3           hPa          VMM
#> 4           hPa          VMM
#> 5           hPa          VMM
#> 6           hPa          VMM
#> 7           hPa          VMM
#> 8           hPa          VMM
```

Each of the stations in the list for a given variable, are represented
by a `ts_id`. These can be used to download the data of a given period
with the command `get_timeseries_tsid`, for example Overpelt (`ts_id
= 78124042`):

``` r
overpelt_pressure <- get_timeseries_tsid("78124042", 
                                         from = "2017-04-01", 
                                         to = "2017-04-02")
head(overpelt_pressure)
#>             Timestamp  Value Quality Code
#> 1 2017-04-01 00:00:00 1008.8          130
#> 2 2017-04-01 00:15:00 1008.7          130
#> 3 2017-04-01 00:30:00 1008.7          130
#> 4 2017-04-01 00:45:00 1008.6          130
#> 5 2017-04-01 01:00:00 1008.5          130
#> 6 2017-04-01 01:15:00 1008.4          130
```

Making a plot of the data with ggplot:

``` r
library(ggplot2)
ggplot(overpelt_pressure, aes(x = Timestamp, y = Value)) + 
    geom_line() + 
    xlab("") + ylab("hPa") + 
    scale_x_datetime(date_labels = "%H:%M\n%Y-%m-%d", date_breaks = "6 hours")
```

<img src="man/figures/README-plot_pressure-1.png" width="80%" />

Another option is to check the available variables for a given station,
with the function `get_variables`. Let’s consider again Overpelt
(`ME11_002`) and check the first ten available variables at the Overpelt
measurement station:

``` r
vars_overpelt <- get_variables("ME11_002")
head(vars_overpelt, 10)
#>    station_name station_no    ts_id    ts_name parametertype_name
#> 1   Overpelt_ME   ME11_002 96216042  DagTotaal         Noverschot
#> 2   Overpelt_ME   ME11_002 78663042   MaandGem                  U
#> 3   Overpelt_ME   ME11_002 78664042   MaandMax                  U
#> 4   Overpelt_ME   ME11_002 78667042       P.10                  U
#> 5   Overpelt_ME   ME11_002 78654042     DagGem                  U
#> 6   Overpelt_ME   ME11_002 78656042     DagMin                  U
#> 7   Overpelt_ME   ME11_002 78658042 HydJaarMax                  U
#> 8   Overpelt_ME   ME11_002 78668042       P.15                  U
#> 9   Overpelt_ME   ME11_002 78660042 KalJaarGem                  U
#> 10  Overpelt_ME   ME11_002 78661042 KalJaarMax                  U
#>    stationparameter_name
#> 1             Noverschot
#> 2                 WSpeed
#> 3                 WSpeed
#> 4                 WSpeed
#> 5                 WSpeed
#> 6                 WSpeed
#> 7                 WSpeed
#> 8                 WSpeed
#> 9                 WSpeed
#> 10                WSpeed
```

Different pre-calculated variables are already available and a `ts_id`
value is available for each of them to download the corresponding data.
For example, `DagGem` (= daily mean values) of `RH` (= relative
humidity), i.e. `ts_id = 78382042`:

``` r
overpelt_rh_daily <- get_timeseries_tsid("78382042", 
                                         from = "2017-04-01", 
                                         to = "2017-04-30")
head(overpelt_rh_daily)
#>             Timestamp Value Quality Code
#> 1 2017-04-01 23:00:00 80.19          130
#> 2 2017-04-02 23:00:00 89.58          130
#> 3 2017-04-03 23:00:00 79.56          130
#> 4 2017-04-04 23:00:00 84.13          130
#> 5 2017-04-05 23:00:00 84.19          130
#> 6 2017-04-06 23:00:00 82.71          130
```

``` r
ggplot(overpelt_rh_daily, aes(x = Timestamp, y = Value)) + 
    geom_line() + 
    xlab("") + ylab(" RH (%)") + 
    scale_x_datetime(date_labels = "%b-%d\n%Y", date_breaks = "5 days")
```

<img src="man/figures/README-plot_rh-1.png" width="80%" />

Unfortunately, not all variables are documented, for which the check for
the appropriate variable is not (yet) fully supported by the package.

More detailed tutorials are available in the package vignettes\!

## Note on restrictions of the downloads

The amount of data downloaded from waterinfo.be is limited via a credit
system. You do not need to get a token right away to download data. For
limited and irregular downloads, a token will not be required.

When you require more extended data requests, please request a download
token from the waterinfo.be site administrators via the e-mail adress
<hydrometrie@waterinfo.be> with a statement of which data and how
frequently you would like to download data. You will then receive a
client-credit code that can be used to obtain a token that is valid for
24 hours, after which the token can be refreshed with the same
client-credit code.

Get token with client-credit code: (limited client-credit code for
testing
purposes)

``` r
client <- paste0("MzJkY2VlY2UtODI2Yy00Yjk4LTljMmQtYjE2OTc4ZjBjYTZhOjRhZGE4",
                 "NzFhLTk1MjgtNGI0ZC1iZmQ1LWI1NzBjZThmNGQyZA==")
my_token <- get_token(client = client)
print(my_token)
#> Token:
#> eyJhbGciOiJIUzI1NiJ9.eyJqdGkiOiI1NjEwNTk1OS02OTRiLTQwYjEtOTdlYy04YTNhNzc0MjQ0MjUiLCJpYXQiOjE1NDQ3NDExMjEsImlzcyI6Imh0dHA6Ly9sb2NhbGhvc3Q6ODA4MC9LaVdlYlBvcnRhbC9hdXRoIiwiYXVkIjoiMzJkY2VlY2UtODI2Yy00Yjk4LTljMmQtYjE2OTc4ZjBjYTZhIiwiZXhwIjoxNTQ0ODI3NTIxfQ.15M1dSilAl7Ui_59wR5nWI9mQawJXZn6j_NFk_YUXYA
#> 
#> Attributes:
#>  url: http://download.waterinfo.be/kiwis-auth/token
#>  type: Bearer
#>  expires: 2018-12-14 23:45:20 CET
```

Receive information on the validity of the token:

``` r
is.expired(my_token)
#> [1] FALSE
```

Check when the token expires:

``` r
expires.in(my_token)
#> Time difference of 24 hours
```

Use token when retrieving data:

``` r
get_stations(variable_name = "verdamping_monteith", token = my_token)
#>      ts_id station_latitude station_longitude station_id station_no
#> 1 94310042         51.02263          2.970584      12206   ME01_003
#> 2 94544042         51.20300          5.439589      12213   ME11_002
#> 3 94530042         51.16224          4.845708      12212   ME10_011
#> 4 94516042         50.73795          5.141976      12211   ME09_012
#> 5 94502042         50.88663          4.094898      12210   ME07_006
#> 6 94474042         51.24379          4.266912      12208   ME04_001
#> 7 94488042         50.86149          3.411318      12209   ME05_019
#> 8 94460042         51.27226          3.728299      12207   ME03_017
#>              station_name stationparameter_name parametertype_name
#> 1               Zarren_ME                   pET                PET
#> 2             Overpelt_ME                   pET                PET
#> 3            Herentals_ME                   pET                PET
#> 4 Niel-bij-St.-Truiden_ME                   pET                PET
#> 5           Liedekerke_ME                   pET                PET
#> 6              Melsele_ME                   pET                PET
#> 7              Waregem_ME                   pET                PET
#> 8            Boekhoute_ME                   pET                PET
#>   ts_unitsymbol dataprovider
#> 1            mm          VMM
#> 2            mm          VMM
#> 3            mm          VMM
#> 4            mm          VMM
#> 5            mm          VMM
#> 6            mm          VMM
#> 7            mm          VMM
#> 8            mm          VMM
```

## Acknowledgements

This package is just a small wrapper around waterinfo.be to facilitate
researchers and other stakeholders in downloading the data from
[waterinfo.be](http://www.waterinfo.be). The availability of this data
is made possible by *de Vlaamse Milieumaatschappij, Waterbouwkundig
Laboratorium, Maritieme Dienstverlening & Kust, Waterwegen en Zeekanaal
NV en De Scheepvaart NV*.

## Meta

  - We welcome [contributions](.github/CONTRIBUTING.md) including bug
    reports.
  - License: MIT
  - Get citation information for `wateRinfo` in R doing
    `citation("wateRinfo")`.
  - Please note that this project is released with a [Contributor Code
    of Conduct](.github/CODE_OF_CONDUCT.md). By participating in this
    project you agree to abide by its
terms.

[![ropensci\_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)


NEWS.md
wateRinfo 0.3.0.9047 (2018-12-13)
=============================

### NEW FEATURES

* WateRinfo moved to ropensci, thanks to @ldecicco-USGS for the review and @karthik for editing the submission

### MINOR IMPROVEMENTS

* Change the URLs to `https` version, which are now supported by waterinfo.be
* Make sure data.frame content are characters and not factors
* Return supported frequencies as vector instead of single ling character

### DOCUMENTATION FIXES

* Update to usage of pkgdown 1.3.0
* Moved link references from INBO tp ropensci
* Add ropensci footer
* Mark Institute of Nature and Forest Research (INBO) as copyright holder
* Updated logo base color

Notice, we will add an in-development fourth component to the release tag with the development version. For more information, see [package guide](http://r-pkgs.had.co.nz/description.html#version).

wateRinfo 0.2.2 (2018-11-07)
=============================

### CRUCIAL QUICK FIX

* When no `datasource` is added to the waterinfo.be query, the server gives a html/text response containing a tomcat server error messag. This error is now captured properly. As the user normally only uses the wrapped API call functions, there should no difference on the user level.

### DOCUMENTATION FIXES

* Bring documentation up to date with the new `datasource` handling

wateRinfo 0.2.1 (2018-10-17)
=============================

The `datasource` of the non-VMM stations changed recently (our tests on CI broke). Moreover, the 
`ts_identifiers` for these stations changed as well. Both are adaptations on the data source (waterinfo.be) level. 

### CRUCIAL QUICK FIX

* Change the datasource to 4 for non-VMM stations
* Adapt the examples and vignettes to existing `ts_identifiers`

wateRinfo 0.2.0 (2018-10-01)
=============================

### NEW FEATURES

* Better support for changed station identifiers of the waterinfo.be dataproviders 

### MINOR IMPROVEMENTS

* Change print statements for message statements
* Add missing test to check token object type (remark: current missing tests are hard to introduce without having control on the original API behaviour)
* Test files have consistent names
* Update code style using styler

### DOCUMENTATION FIXES

* Describe the table output format explicitly in roxygen docs and add the quality codes of the waterinfo.be dataproviders (#27)
* Update contributing guidelines, thanks to @peterdesmet who did a great effort on providing welcoming, clear written and engaging contributing guidelines
* Provide code examples for all public functions
* Provide top-level documentation handle `?wateRinfo` (#23)
* Add keywords internal of internal functions not part of the documentation index (#24)


wateRinfo 0.1.2 (2018-05-03)
============================

* Added a `NEWS.md` file to track changes to the package (#25).
* Provide support for parsible non-existing dates (e.g. '2018-04-31')
* Add contributing guidelines


wateRinfo 0.1.1 (2017-11-29)
============================

* Add support for token handling as distributed by waterinfo.be





R/get_timeseries.R

#' Download timeseries data from waterinfo.be
#'
#' Using the ts_id codes  and by providing a given date period, download the
#' corresponding time series from the waterinfo.be website
#'
#'
#' @param ts_id waterinfo.be database ts_id, defining a timeserie variable and
#' frequency it is defined.
#' @param period input string according to format required by waterinfo:
#' De period string is provided as P#Y#M#DT#H#M#S, with P defines `Period`,
#' each # is an integer value and the codes define the number of...
#' Y - years
#' M - months
#' D - days
#' T required if information about sub-day resolution is present
#' H - hours
#' D - days
#' M - minutes
#' S - seconds
#' Instead of D (days), the usage of W - weeks is possible as well
#' Examples of valid period strings: P3D, P1Y, P1DT12H, PT6H, P1Y6M3DT4H20M30S.
#' @param from date of datestring as start of the time series
#' @param to date of datestring as end of the time series
#' @param datasource int [0-3] defines the `meetnet` of which the measurement
#' station is part of. VMM based stations are net '1', MOW-HIC is net '2'
#' @param token token to use with the call (optional, can be retrieved via
#' \code{\link{get_token}})
#'
#' @return data.frame with the timestamps, values and quality code
#'
#' @format A data.frame with 3 variables:
#' \describe{
#'   \item{Timestamp}{Datetime of the measurement.}
#'   \item{Value}{Measured value.}
#'   \item{Quality Code}{Quality code of the measurement, dependent on the
#'   data source used:
#'      \itemize{
#'         \item{VMM Quality Code Interpretation (datasource 1)
#'            \itemize{
#'               \item{10/110 - Excellent}
#'               \item{30/100/130 - Good}
#'               \item{50/150 - Moderate}
#'               \item{70/170 - Poor}
#'               \item{80/180 - Estimated}
#'               \item{90/190 - Suspect}
#'               \item{220 - Default}
#'               \item{-1 - Missing}
#'            }
#'         }
#'         \item{HIC Quality Code Interpretation (datasource 2)
#'            \itemize{
#'               \item{40 - Good}
#'               \item{80 - Estimated}
#'               \item{120 - Suspect}
#'               \item{200 - Unchecked}
#'               \item{60 - Complete}
#'               \item{160 - Incomplete}
#'               \item{-1 - Missing}
#'            }
#'         }
#'         \item{Aggregated timeseries
#'            \itemize{
#'               \item{40 - Good}
#'               \item{100 - Estimated}
#'               \item{120 - Suspect}
#'               \item{200 - Unchecked}
#'               \item{-1 - Missing}
#'            }
#'         }
#'      }
#'    }
#' }
#' The URL of the specific request is provided as a comment attribute to the
#' returned data.frame. Use \code{comment(df)} to get the request URL.
#'
#'
#' @export
#' @importFrom lubridate ymd_hms
#'
#' @examples
#' get_timeseries_tsid("35055042", from = "2017-01-01", to = "2017-01-02")
#' get_timeseries_tsid("5156042", period = "P3D")
#' get_timeseries_tsid("55419010", from = "2017-06-01", to = "2017-06-03",
#'                     datasource = 4)
get_timeseries_tsid <- function(ts_id, period = NULL, from = NULL,
                                to = NULL, datasource = 1, token = NULL) {
  # define the date fields we require
  return_fields <- c("Timestamp", "Value", "Quality Code")

  # check and handle the date/period information
  period_info <- parse_period(from, to, period)

  # general arguments
  query_list <- list(
    type = "queryServices", service = "kisters",
    request = "getTimeseriesvalues",
    ts_id = ts_id, format = "json",
    datasource = datasource,
    returnfields = as.character(paste(return_fields,
      collapse = ","
    ))
  )

  # http GET call to waterinfo for the dataframe
  time_series <- call_waterinfo(
    query = c(query_list, period_info),
    token = token
  )

  if (time_series$content$rows == 0) {
    df <- data.frame(
      Timestamp = as.POSIXct(character()),
      Value = double(),
      "Quality Code" = character(),
      stringsAsFactors = FALSE
    )
    colnames(df) <- return_fields
  } else {
    df <- as.data.frame(time_series$content$data,
      stringsAsFactors = FALSE
    )

    # data formatting:
    df$X1 <- ymd_hms(df$X1)
    df$X2 <- as.double(as.character(df$X2))
    colnames(df) <- strsplit(time_series$content$columns, ",")[[1]]
  }

  # add request URL as df comment
  comment(df) <- time_series$response$url

  return(df)
}
